### PR TITLE
Added choosable formatting filesystem for LVs

### PR DESCRIPTION
**Pull request by mickydelfavero**
LXD has hardcoded ext4 as the filesystem to be used when LVM storage backend is choosed, I don't like very much ext* filesystems and prefer xfs over them so I've done a little patch to be able to use it with LXD.

To be able to choose preferred filesystem a key was added to the configuration parameter of LXD: storage.lvm_fstype, for now it's value can be only ext4 (default, so I don't break anything) and xfs, but it's trivial to add other filesystems.

For now the filesystem has to be choosen at LXD installation time, in the future I think it will be better to let the users to choose filesystem at container creation time so it'll be possible to have different filesystem for different container and so migrating a container from one host to another will be possible even if they have different default filesystem.

### CODE REVIEW

**Comment by stgraber** [comment]
Hi, 

Can you sign-off your commit? (see CONTRIBUTING.md)

Also, why is it necessary to call xfs_admin all the time for that uuid? isn't that done for you at mount time? 
If it is, then you could use the generic mounting function from lxd/devices which detects the filesystem type for you.

**Comment by mickydelfavero** [comment]
Hi, Stéphane, thank for your comment.

Generating a new xfs'UUID when creating a container, an image or a snapshot is necessary because isn't possible to mount two xfs volumes with the same UUID.

    root@lina:~# xfs_admin -u /dev/lxc/w3
    UUID = 834ea86b-f8a8-4076-b04d-16f0b3d4a898
    root@lina:~# xfs_admin -u /dev/lxc/w4
    UUID = 834ea86b-f8a8-4076-b04d-16f0b3d4a898
    root@lina:~# mount /dev/lxc/w3 /tmp/p
    root@lina:~# mount /dev/lxc/w4 /tmp/q
    mount: wrong fs type, bad option, bad superblock on /dev/mapper/lxc-w4,
               missing codepage or helper program, or other error

           In some cases useful info is found in syslog - try
           dmesg | tail or so.
    root@lina:~# dmesg | tail -1 
    [1116904.345407] XFS (dm-11): Filesystem has duplicate UUID 834ea86b-f8a8-4076-b04d-16f0b3d4a898 - can't mount
    root@lina:~#

File: lxd/storage_lvm.go [hunk 1]
```diff
@@ -763,7 +813,25 @@ func (s *storageLvm) ImageCreate(fingerprint string) error {
...
+	if fstype == "xfs" {
+		output, err := exec.Command(
+			"xfs_admin",
+			"-U", "generate",
+			lvpath).CombinedOutput()
+		if err != nil {
+			// s.ContainerDelete(container)
+			return fmt.Errorf("Error generating new UUID: %v\noutput:'%s'", err, string(output))
+		}
+	}
```
Comment (stgraber) [inline lxd/storage_lvm.go]: I don't see why we need to generate a new UUID here. We just got a new clean xfs filesystem created through mkfs.xfs, it's not a clone of any existing filesystem, so it should have a unique UUID.

**Comment by stgraber** [comment]
Okay, I guess that makes sense. I see two cases where it's indeed needed (new container from an image and mounting a snapshot), it's however not needed when creating a new image (commented).

I'd also prefer if you made the xfs uuid re-generation a separate function which you call from the two places where it's needed.

Thanks!

**Comment by stgraber** [comment]
Oh and can you also update specs/configuration.md with the new config key?

**Comment by mickydelfavero** [comment]
Hi Stéfan, I've made a commit with your suggestions.

File: specs/configuration.md [hunk 2]
```diff
@@ -23,6 +23,7 @@ core.https\_address             | string        | -                         | Ad
 core.trust\_password            | string        | -                         | Password to be provided by clients to setup a trust
 storage.lvm\_vg\_name           | string        | -                         | LVM Volume Group name to be used for container and image storage. A default Thin Pool is created using 100% of the free space in the Volume Group, unless `storage.lvm_thinpool_name` is set.
 storage.lvm\_thinpool\_name     | string        | "LXDPool"                 | LVM Thin Pool to use within the Volume Group specified in `storage.lvm_vg_name`, if the default pool parameters are undesirable.
+storage.lvm\_fstype             | string        | -                         | Format LV with filesystem, for now it's value can be only ext4 (default) or xfs.
```
Comment (stgraber) [inline specs/configuration.md]: Should say "ext4" instead of "-" in the default value column.

**Comment by stgraber** [comment]
Cool, just one tiny change to specs/configuration.md and we'll be good to test and merge this!

**Comment by mickydelfavero** [comment]
I changed it.

**Comment by stgraber** [comment]
Great, lets throw that at Jenkins.

jenkins: test this please

**Comment by stgraber** [comment]
Looks like all the LVM tests are failing, so there's something wrong with this pull request, looking at the diff again...

File: lxd/storage_lvm.go [hunk 3]
```diff
@@ -763,7 +819,13 @@ func (s *storageLvm) ImageCreate(fingerprint string) error {
 		}
 	}()
 
-	err = s.tryMount(lvpath, tempLVMountPoint, "ext4", 0, "discard")
+	var fstype string
+	fstype, err = s.d.ConfigValueGet("storage.lvm_fstype")
+	if err != nil {
+		return fmt.Errorf("Error checking server config, err=%v", err)
+	}
```
Comment (stgraber) [inline lxd/storage_lvm.go]: Needs to set to ext4 if value is empty.

File: lxd/storage_lvm.go [hunk 4]
```diff
@@ -428,7 +464,12 @@ func (s *storageLvm) ContainerCopy(container container, sourceContainer containe
 func (s *storageLvm) ContainerStart(container container) error {
 	lvName := containerNameToLVName(container.Name())
 	lvpath := fmt.Sprintf("/dev/%s/%s", s.vgName, lvName)
-	err := s.tryMount(lvpath, container.Path(), "ext4", 0, "discard")
+	fstype, err := s.d.ConfigValueGet("storage.lvm_fstype")
+	if err != nil {
+		return fmt.Errorf("Error checking server config, err=%v", err)
+	}
```
Comment (stgraber) [inline lxd/storage_lvm.go]: Needs to set to ext4 if value is empty.

File: lxd/storage_lvm.go [hunk 5]
```diff
@@ -709,7 +750,22 @@ func (s *storageLvm) ContainerSnapshotStart(container container) error {
 		}
 	}
 
-	err = s.tryMount(lvpath, container.Path(), "ext4", 0, "discard")
+	var fstype string
+	fstype, err = s.d.ConfigValueGet("storage.lvm_fstype")
+	if err != nil {
+		return fmt.Errorf("Error checking server config, err=%v", err)
+	}
```
Comment (stgraber) [inline lxd/storage_lvm.go]: Needs to set to ext4 if value is empty.

File: lxd/storage_lvm.go [hunk 6]
```diff
@@ -334,7 +355,22 @@ func (s *storageLvm) ContainerCreateFromImage(
 		return err
 	}
 
-	err = s.tryMount(lvpath, destPath, "ext4", 0, "discard")
+	var fstype string
+	fstype, err = s.d.ConfigValueGet("storage.lvm_fstype")
+	if err != nil {
+		return fmt.Errorf("Error checking server config, err=%v", err)
+	}
```
Comment (stgraber) [inline lxd/storage_lvm.go]: Needs to set to ext4 if value is empty.

**Comment by stgraber** [comment]
Oops, the problem was pretty obvious, I should have thought of it on first review :)

**Comment by mickydelfavero** [comment]
Stupid me that I didn't do it on first commit...

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by mikemccracken** [comment]
Hi, I didn't have time for a full review, but wanted to leave a drive-by note about FS features required by LXD/LVM:

xfs should be fine in terms of support for `fs_freeze`, which is necessary to make the snapshotting and copying work safely.

However, it might be worth looking into whether xfs has any similar issues to what we found with ext4, where the first snapshot was the expected small size, but subsequent snapshots of the same origin LV were larger and larger. In LXD terms, this meant that the first container you created from an image used a tiny amount of additional space, but subsequent containers created from the same image would start getting bigger, counter to what you'd expect from a COW snapshot.
This was apparently due to some background initialization (see the `lazy_itable_init` etc in the mkfs.ext4 call) that meant that the origin LV was still changing as the initial containers were being created, causing inefficient allocation.

We noticed this by tracking usage as we created a large number of containers with the LVM backend, using basically the https://github.com/lxc/lxd/blob/master/test/extras/speedtest_create.sh script.

**Pull request merged**